File: common/src/main/java/me/nickimpact/gts/storage/implementation/file/ConfigurateStorage.java
/*
 * This file is part of LuckPerms, licensed under the MIT License.
 *
 *  Copyright (c) lucko (Luck) <luck@lucko.me>
 *  Copyright (c) contributors
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in all
 *  copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 */

package me.nickimpact.gts.storage.implementation.file;

import com.google.common.base.Throwables;
import me.nickimpact.gts.api.listings.Listing;
import me.nickimpact.gts.api.listings.SoldListing;
import me.nickimpact.gts.api.plugin.IGTSPlugin;
import me.nickimpact.gts.storage.implementation.StorageImplementation;
import me.nickimpact.gts.storage.implementation.file.loaders.ConfigurateLoader;
import ninja.leaping.configurate.ConfigurationNode;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.UUID;

public class ConfigurateStorage implements StorageImplementation {

    private final IGTSPlugin plugin;
    private final String implementationName;

    // The loader responsible for I/O
    private final ConfigurateLoader loader;

    private String extension;

    private Path dataDir;
    private String dataDirName;

    private Path userDir;

    private FileWatcher.WatchedLocation userWatcher = null;

    public ConfigurateStorage(IGTSPlugin plugin, String implementationName, ConfigurateLoader loader, String extension, String dataDirName) {
        this.plugin = plugin;
        this.implementationName = implementationName;
        this.loader = loader;
        this.extension = extension;
        this.dataDirName = dataDirName;
    }

    @Override
    public IGTSPlugin getPlugin() {
        return this.plugin;
    }

    @Override
    public String getName() {
        return this.implementationName;
    }

    @Override
    public void init() throws Exception {
        this.dataDir = this.plugin.getConfigDir().resolve(this.dataDirName);
        this.createDirectoriesIfNotExists(this.dataDir);

        this.userDir = this.dataDir.resolve("data");
        this.createDirectoriesIfNotExists(this.userDir);

        FileWatcher watcher = new FileWatcher(this.plugin, dataDir);
        this.userWatcher = watcher.getWatcher(this.userDir);
        this.userWatcher.addListener(path -> {
//            String s = path.getFileName().toString();
//
//            if (!s.endsWith(this.extension)) {
//                return;
//            }
//
//            String user = s.substring(0, s.length() - this.extension.length());
//            UUID uuid;
//            try {
//                uuid = UUID.fromString(user);
//            } catch (Exception e) {
//                return;
//            }
        });
    }

    @Override
    public void shutdown() throws Exception {}

    @Override
    public boolean addListing(Listing listing) throws Exception {
        return false;
    }

    @Override
    public boolean deleteListing(UUID uuid) throws Exception {
        return false;
    }

    @Override
    public List<Listing> getListings() throws Exception {
        return null;
    }

    @Override
    public boolean addIgnorer(UUID uuid) throws Exception {
        return false;
    }

    @Override
    public boolean removeIgnorer(UUID uuid) throws Exception {
        return false;
    }

    @Override
    public List<UUID> getAllIgnorers() throws Exception {
        return null;
    }

    @Override
    public boolean addToSoldListings(UUID owner, SoldListing listing) {
        return false;
    }

    @Override
    public List<SoldListing> getAllSoldListingsForPlayer(UUID uuid) {
        return null;
    }

    @Override
    public boolean deleteSoldListing(UUID id, UUID owner) throws Exception {
        return false;
    }

    @Override
    public boolean purge() throws Exception {
        return false;
    }

    private ConfigurationNode readFile(String name) throws IOException {
        Path file = this.userDir.resolve(name + this.extension);
        if(this.userWatcher != null) {
            this.userWatcher.recordChange(file.getFileName().toString());
        }

        if(!Files.exists(file)) {
            return null;
        }

        return this.loader.loader(file).load();
    }

    private void saveFile(String name, ConfigurationNode node) throws IOException {
        Path file = this.userDir.resolve(name + this.extension);
        if(this.userWatcher != null) {
            this.userWatcher.recordChange(file.getFileName().toString());
        }

        if(node == null) {
            Files.deleteIfExists(file);
            return;
        }

        this.loader.loader(file).save(node);
    }

    private void createDirectoriesIfNotExists(Path path) throws IOException {
        if (Files.exists(path) && (Files.isDirectory(path) || Files.isSymbolicLink(path))) {
            return;
        }

        Files.createDirectories(path);
    }

    // used to report i/o exceptions which took place in a specific file
    private RuntimeException reportException(String file, Exception ex) throws RuntimeException {
        this.plugin.getPluginLogger().warn("Exception thrown whilst performing i/o: " + file);
        ex.printStackTrace();
        throw Throwables.propagate(ex);
    }
}

File: common/src/main/kotlin/net/impactdev/gts/common/storage/implementation/file/ConfigurateStorage.kt
/*
 * This file is part of LuckPerms, licensed under the MIT License.
 *
 *  Copyright (c) lucko (Luck) <luck@lucko.me>
 *  Copyright (c) contributors
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in all
 *  copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 */
package net.impactdev.gts.common.storage.implementation.file

import com.github.benmanes.caffeine.cache.Caffeine
import com.github.benmanes.caffeine.cache.LoadingCache
import com.google.common.base.Throwables
import com.google.common.collect.ImmutableMap
import net.impactdev.gts.api.listings.Listing
import net.impactdev.gts.api.messaging.message.type.admin.ForceDeleteMessage
import net.impactdev.gts.api.messaging.message.type.auctions.AuctionMessage
import net.impactdev.gts.api.messaging.message.type.listings.BuyItNowMessage
import net.impactdev.gts.api.messaging.message.type.listings.BuyItNowMessage.Purchase
import net.impactdev.gts.api.messaging.message.type.listings.ClaimMessage
import net.impactdev.gts.api.player.PlayerSettings
import net.impactdev.gts.api.stashes.Stash
import net.impactdev.gts.common.plugin.GTSPlugin
import net.impactdev.gts.common.storage.implementation.StorageImplementation
import net.impactdev.gts.common.storage.implementation.file.loaders.ConfigurateLoader
import net.impactdev.gts.common.utils.exceptions.ExceptionWriter
import ninja.leaping.configurate.ConfigurationNode
import ninja.leaping.configurate.SimpleConfigurationNode
import java.io.IOException
import java.nio.file.*
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.locks.ReentrantLock
import java.util.function.Consumer
import java.util.function.Function
import kotlin.collections.List
import kotlin.collections.Map
import kotlin.collections.set

class ConfigurateStorage(
    override val plugin: GTSPlugin, private val implementationName: String, // The loader responsible for I/O
    private val loader: ConfigurateLoader, private val extension: String, private val dataDirName: String
) : StorageImplementation {
    private var dataDir: Path? = null
    private val fileGroups: Map<Group, FileGroup>
    private val users: FileGroup
    private override val listings: FileGroup
    private val watcher: FileWatcher?

    private enum class Group {
        USERS, LISTINGS
    }

    private class FileGroup {
        val directory: Path? = null
        val watcher: FileWatcher.WatchedLocation? = null
    }

    private val ioLocks: LoadingCache<Path, ReentrantLock>
    override val name: String?
        get() = implementationName

    @kotlin.Throws(Exception::class)
    override fun init() {
        dataDir = Paths.get("gts")
        createDirectoriesIfNotExists(dataDir)
        users.directory = dataDir.resolve("users")
        listings.directory = dataDir.resolve("listings")
        val uuidParser = label@ Function<String, UUID> { input: String? ->
            try {
                return@label UUID.fromString(input)
            } catch (e: IllegalArgumentException) {
                return@label null
            }
        }
        if (watcher != null) {
            users.watcher = watcher.getWatcher(users.directory)
            users.watcher.addListener(Consumer { path: Path ->
                val file = path.fileName.toString()
                if (!file.endsWith(this.extension)) {
                    return@addListener
                }
                val user = file.substring(0, file.length - this.extension.length)
                val id = uuidParser.apply(user) ?: return@addListener
                val name: String = GTSPlugin.Companion.getInstance().getPlayerDisplayName(id)
                plugin.pluginLogger.info("[File Watcher] Detected change in user file for $name")
            })
            listings.watcher = watcher.getWatcher(listings.directory)
            listings.watcher.addListener(Consumer { path: Path ->
                val file = path.fileName.toString()
                if (!file.endsWith(this.extension)) {
                    return@addListener
                }
                val user = file.substring(0, file.length - this.extension.length)
                val id = uuidParser.apply(user) ?: return@addListener
                plugin.pluginLogger.info("[File Watcher] Detected change in listing file with ID: $id")
            })
        }
    }

    @kotlin.Throws(Exception::class)
    override fun shutdown() {
    }

    @kotlin.Throws(Exception::class)
    override fun addListing(listing: Listing?): Boolean {
        try {
            val file: ConfigurationNode = SimpleConfigurationNode.root()
            file.getNode("data").value = listing!!.serialize()!!.toJson()
        } catch (e: Exception) {
            return false
        }
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun deleteListing(uuid: UUID?): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun getListing(id: UUID?): Optional<Listing> {
        return Optional.empty()
    }

    @kotlin.Throws(Exception::class)
    override fun getListings(): List<Listing?>? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun hasMaxListings(user: UUID?): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun purge(): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun clean(): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun getStash(user: UUID?): Stash? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun getPlayerSettings(user: UUID?): Optional<PlayerSettings> {
        return Optional.empty()
    }

    @kotlin.Throws(Exception::class)
    override fun applyPlayerSettings(user: UUID?, updates: PlayerSettings?): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun processPurchase(request: Purchase.Request?): Purchase.Response? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun sendListingUpdate(listing: Listing): Boolean {
        return false
    }

    override fun processBid(request: AuctionMessage.Bid.Request?): AuctionMessage.Bid.Response? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun processClaimRequest(request: ClaimMessage.Request?): ClaimMessage.Response? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun appendOldClaimStatus(auction: UUID?, lister: Boolean, winner: Boolean, others: List<UUID?>?): Boolean {
        return false
    }

    @kotlin.Throws(Exception::class)
    override fun processAuctionCancelRequest(request: AuctionMessage.Cancel.Request?): AuctionMessage.Cancel.Response? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun processListingRemoveRequest(request: BuyItNowMessage.Remove.Request?): BuyItNowMessage.Remove.Response? {
        return null
    }

    @kotlin.Throws(Exception::class)
    override fun processForcedDeletion(request: ForceDeleteMessage.Request?): ForceDeleteMessage.Response? {
        return null
    }

    @kotlin.Throws(IOException::class)
    private fun readFile(uuid: UUID): ConfigurationNode? {
        //Path file = this.getDirectory()
        return null
    }

    @kotlin.Throws(IOException::class)
    private fun saveFile(name: String, node: ConfigurationNode) {
    }

    @kotlin.Throws(IOException::class)
    private fun createDirectoriesIfNotExists(path: Path?) {
        if (Files.exists(path) && (Files.isDirectory(path) || Files.isSymbolicLink(path))) {
            return
        }
        Files.createDirectories(path)
    }

    // used to report i/o exceptions which took place in a specific file
    @kotlin.Throws(RuntimeException::class)
    private fun reportException(file: String, ex: Exception): RuntimeException {
        plugin.pluginLogger.warn("Exception thrown whilst performing i/o: $file")
        ex.printStackTrace()
        throw Throwables.propagate(ex)
    }

    init {
        users = FileGroup()
        listings = FileGroup()
        val fileGroups = EnumMap<Group, FileGroup>(
            Group::class.java
        )
        fileGroups[Group.USERS] = users
        fileGroups[Group.LISTINGS] = listings
        this.fileGroups = ImmutableMap.copyOf(fileGroups)
        val watcher: FileWatcher?
        watcher = try {
            FileWatcher(Paths.get("gts"), true)
        } catch (e: Throwable) {
            GTSPlugin.Companion.getInstance().getPluginLogger()
                .error("Error occurred whilst trying to create a file watcher...")
            ExceptionWriter.write(e)
            null
        }
        this.watcher = watcher
        ioLocks = Caffeine.newBuilder()
            .expireAfterAccess(10, TimeUnit.MINUTES)
            .build { key: Path? -> ReentrantLock() }
    }
}